### PR TITLE
Limit max call frame overhead.

### PR DESCRIPTION
**Pull request by Raynos**
This PR implements #486

 - Max arg1 is 16kb (already in master)
 - Max transport header count is 128 (this PR)
 - Max transport header key length is 16 (this PR)

In this PR we also improve how we handle protocol read and write
failures as seen in the new `sendProtocolError()` method in
connection.js

r: @jcorbin @rf

### CODE REVIEW

File: docs/protocol.md [hunk 1]
```diff
@@ -307,6 +307,8 @@ The meaning of the three args depends on the systems on each end. The format of
 arg1, arg2, and arg3 is unspecified at the transport level. These are opaque
 binary blobs as far as tchannel is concerned.
 
+The size of `arg1` is at most 16 kilobytes.
```
Comment (jcorbin) [inline docs/protocol.md]: Please spell this `16KiB`.

The phrase "16 kilobytes" is open to interpretation, but "16KiB" means `16 * 1024 = 16384`

File: node/connection.js [hunk 2]
```diff
@@ -90,16 +90,9 @@ TChannelConnection.prototype.setupSocket = function setupSocket() {
     self.socket.on('error', onSocketError);
 
     function onSocketChunk(chunk) {
-        self.mach.handleChunk(chunk, chunkHandled);
-    }
-
-    function chunkHandled(err) {
+        var err = self.mach.handleChunk(chunk);
```
Comment (jcorbin) [inline node/connection.js]: nice fix... could even make this a separate pr if this one takes long-ish

File: node/connection.js [hunk 3]
```diff
@@ -191,6 +184,41 @@ TChannelConnection.prototype.setupHandler = function setupHandler() {
...
+function sendProtocolError(type, err) {
+    var self = this;
+
+    assert(type === 'write' || type === 'read',
+        'Got invalid type: ' + type);
+
+    var protocolError;
+
+    if (type === 'read') {
+        protocolError = errors.TChannelReadProtocolError(err, {
```
Comment (jcorbin) [inline node/connection.js]: don't see this defined earlier... is this path even tested?

File: node/connection.js [hunk 4]
```diff
@@ -191,6 +184,41 @@ TChannelConnection.prototype.setupHandler = function setupHandler() {
...
+        });
+
+        self.handler.sendErrorFrame({
+            id: protocolError.frameId || 0xFFFFFFFF
+        }, 'ProtocolError', protocolError.message);
+
+        self.resetAll(protocolError);
+        self.socket.destroy();
+    } else if (type === 'write') {
+        protocolError = errors.TChannelWriteProtocolError(err, {
```
Comment (jcorbin) [inline node/connection.js]: don't see this defined earlier... is this path even tested?

File: node/errors.js [hunk 5]
```diff
@@ -372,6 +372,13 @@ module.exports.TChannelLocalResetError = WrappedError({
     message: 'tchannel: {causeMessage}'
 });
 
+module.exports.TChannelProtocolError = WrappedError({
```
Comment (jcorbin) [inline node/errors.js]: neither read nor write... huh?

File: node/v2/header.js [hunk 6]
```diff
@@ -28,14 +28,16 @@ var errors = require('../errors');
...
+function HeaderRW(countrw, keyrw, valrw, options) {
     if (!(this instanceof HeaderRW)) {
-        return new HeaderRW(countrw, keyrw, valrw);
+        return new HeaderRW(countrw, keyrw, valrw, options);
     }
     var self = this;
     self.countrw = countrw;
     self.keyrw = keyrw;
     self.valrw = valrw;
+    self.enforceHeaderCount = options.enforceHeaderCount;
```
Comment (jcorbin) [inline node/v2/header.js]: can we just make this:
```
options.maxHeaderCount :: Number
options.maxKeyLength :: Number
```

Rather than booleans and hardcoded numbers?

File: node/v2/header.js [hunk 7]
```diff
@@ -69,11 +71,28 @@ HeaderRW.prototype.writeInto = function writeInto(headers, buffer, offset) {
...
+            count: keys.length
+        }), offset);
+    }
+
     for (var i = 0; i < keys.length; i++) {
         if (res.err) return res;
         offset = res.offset;
 
         var key = keys[i];
+        // TODO: Check that its' 16 bytes
```
Comment (jcorbin) [inline node/v2/header.js]: isn't this TO-DONE?

File: node/v2/header.js [hunk 8]
```diff
@@ -69,11 +71,28 @@ HeaderRW.prototype.writeInto = function writeInto(headers, buffer, offset) {
...
+            return bufrw.WriteResult.error(errors.TransportHeaderTooLong({
+                offset: offset,
+                endOffset: res.offset,
+                headerName: key
+            }), offset);
+        }
+
         res = self.keyrw.writeInto(key, buffer, offset);
         if (res.err) return res;
         offset = res.offset;
```
Comment (jcorbin) [inline node/v2/header.js]: you can TO-DONE that by keeping a start offset, and then calculating the difference here.

Basically: just do the write, and then check the bytes here

File: node/v2/header.js [hunk 9]
```diff
@@ -110,6 +137,13 @@ HeaderRW.prototype.readFrom = function readFrom(buffer, offset) {
...
             }), offset, headers);
+        // TODO: check key is 16 bytes; not 16 characters
+        } else if (self.enforceKeyLength && key.length > 16) {
+            return bufrw.ReadResult.error(errors.TransportHeaderTooLong({
+                offset: offset,
+                endOffset: res.offset,
+                headerName: key
+            }), offset, headers);
         }
         offset = res.offset;
```
Comment (jcorbin) [inline node/v2/header.js]: likewise

File: node/errors.js [hunk 10]
```diff
@@ -452,6 +467,14 @@ module.exports.TopLevelRequestError = TypedError({
         'Must use a sub channel directly.'
 });
 
+module.exports.TransportHeaderTooLong = TypedError({
+    type: 'tchannel.tranport-header-too-long',
```
Comment (rf) [inline node/errors.js]: tranport

File: node/v2/header.js (hunk 6 above)
Comment (rf) [inline node/v2/header.js]: :+1:

**Comment by Raynos** [comment]
Addressed all feedback. cc @rf @jcorbin

**Comment by Raynos** [comment]
Travis passed. #fierce

**Comment by rf** [comment]
:+1:
